stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,300 @@
+      return false;
+    }
+
+    // Validate the binary is current ELF executable for the CPU
+    // bitsize
+    if (elfHeader->e_ident[EI_CLASS] != ELFCLASS
+        || elfHeader->e_type != ET_EXEC
+        || elfHeader->e_version != EV_CURRENT
+        || elfHeader->e_ehsize != sizeof(Elf_Ehdr)) {
+      return false;
+    }
+
+    // Map in the program header block
+    auto programHeaderSize = (elfHeader->e_phentsize * elfHeader->e_phnum);
+    if (!expandMapping(elfHeader->e_phoff + programHeaderSize)) {
+      return false;
+    }
+
+    auto headers = reinterpret_cast<const Elf_Phdr *>(mapping +
+                                                      elfHeader->e_phoff);
+    // If a interpreter is set in the program headers then this is a
+    // dynamic executable and therefore not valid
+    for (size_t idx = 0; idx < elfHeader->e_phnum; idx++) {
+      if (headers[idx].p_type == PT_INTERP) {
+        return false;
+      }
+    }
+
+    // Map in the section headers
+    auto sectionHeaderSize = (elfHeader->e_shentsize * elfHeader->e_shnum);
+    if (!expandMapping(elfHeader->e_shoff + sectionHeaderSize)) {
+      return false;
+    }
+    symbolTable = findSection(SHT_SYMTAB);
+    stringTable = findSection(SHT_STRTAB);
+
+    return true;
+  }
+
+  // Find the section of a specified type expanding the mapping as needed
+  const Elf_Shdr *findSection(Elf_Word sectionType) {
+    auto headers = reinterpret_cast<const Elf_Shdr *>(mapping +
+                                                      elfHeader->e_shoff);
+
+    for (size_t idx = 0; idx < elfHeader->e_shnum; idx++) {
+      if (idx == elfHeader->e_shstrndx) {
+        continue;
+      }
+      auto header = &headers[idx];
+      if (header->sh_type == sectionType) {
+        if (header->sh_entsize > 0 && header->sh_size % header->sh_entsize) {
+          fprintf(stderr,
+                  "section size is not a multiple of entrysize (%ld/%ld)\n",
+                  header->sh_size, header->sh_entsize);
+          return nullptr;
+        }
+
+        if (!expandMapping(header->sh_offset + header->sh_size)) {
+          return nullptr;
+        }
+        return header;
+      }
+    }
+    return nullptr;
+  }
+
+  bool expandMapping(size_t size) {
+    if (mapping == nullptr || size > fileSize) {
+      return false;
+    }
+    if (size > mapLength) {
+      void *map = mremap(const_cast<char *>(mapping), mapLength, size,
+                         MREMAP_MAYMOVE);
+      if (map == MAP_FAILED) {
+        return false;
+      }
+      mapLength = size;
+      mapping = reinterpret_cast<const char *>(map);
+      elfHeader = reinterpret_cast<const Elf_Ehdr *>(map);
+    }
+    return true;
+  }
+};
+
+int
+swift::lookupSymbol(const void *address, SymbolInfo *info) {
+  // The pointers returned point into the mmap()'d binary so keep the
+  // object once instantiated.
+  static auto binary = new StaticBinaryELF(binaryFilename);
```

Comment:
It shouldn't be necessary to allocate anything at all if the lookup is done in the copy of the executable already mapped by the kernel.